data_model/src/events/time.rs

```diff
@@ -0,0 +1,390 @@
+//! Time event and filter
+
+use core::{cmp::Ordering, ops::Range, time::Duration};
+
+use super::*;
+
+/// Special event that is emitted when `WSV` is ready for handling time-triggers
+///
+/// Contains time interval which is used to identify time-triggers to be executed
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Decode, Encode, IntoSchema)]
+pub struct Event {
+    /// Previous block timestamp and consensus durations estimation.
+    /// `None` if it's first block commit
+    pub prev_interval: Option<Interval>,
+    /// Current block timestamp and consensus durations estimation
+    pub interval: Interval,
+}
+
+impl Event {
+    /// Construct `Event` with `prev_interval` and `interval`
+    pub fn new(prev_interval: Option<Interval>, interval: Interval) -> Self {
+        Self {
+            prev_interval,
+            interval,
+        }
+    }
+}
+
+/// Filters time-events and allows only ones which time interval contains
+#[derive(
+    Debug,
+    PartialOrd,
+    Ord,
+    PartialEq,
+    Eq,
+    Clone,
+    Copy,
+    Decode,
+    Encode,
+    IntoSchema,
+    Hash,
+    Serialize,
+    Deserialize,
+)]
+pub struct EventFilter(pub Reoccurs);
+
+impl EventFilter {
+    /// Compute how much times trigger with `self` as filter should be executed on `event`
+    pub fn count_matches(&self, event: &Event) -> u32 {
+        match &self.0 {
+            Reoccurs::Periodically(periodicity) => {
+                let mut res = Self::count_matches_in_interval(periodicity, &event.interval);
+
+                if let Some(prev_interval) = event.prev_interval {
+                    let previous_estimation = prev_interval.since + prev_interval.length;
+
+                    // Counting matching points between estimated timestamp and the real one
+                    match event.interval.since.cmp(&previous_estimation) {
+                        Ordering::Greater => {
+                            let length = event.interval.since - previous_estimation;
+                            let forgotten_count = Self::count_matches_in_interval(
+                                periodicity,
+                                &Interval::new(previous_estimation, length),
+                            );
+                            res += forgotten_count;
+                        }
+                        Ordering::Less => {
+                            let length = previous_estimation - event.interval.since;
+                            let twice_counted_count = Self::count_matches_in_interval(
+                                periodicity,
+                                &Interval::new(event.interval.since, length),
+                            );
+                            res -= twice_counted_count;
+                        }
+                        Ordering::Equal => (),
+                    }
+                }
+
+                res
+            }
```

Comment:
```suggestion
            Reoccurs::Periodically(periodicity) => {
                let between_estimations = event.prev_interval.map_or(event.interval, |prev| {
                    let estimation = event.interval.since + event.interval.length;
                    let prev_estimation = prev.since + prev.length;
                    Interval::new(prev_estimation, estimation.saturating_sub(prev_estimation))
                });
                Self::count_matches_in_interval(periodicity, &between_estimations)
            },
```

Reply:
Again much more clean than mine, thank you!